### PR TITLE
fix: crash when session expires FS-1666

### PR DESCRIPTION
**Pull request by johnxnguyen**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

If the session expires (e.g the self client is deleted remotely), then the app will crash.

### Causes

When the self client is deleted, we try to delete the keystore, even if proteus via core crypto is enabled. The keystore doesn't exist, resulting in a crash.

### Solutions

Use `ProteusProvider` to access the keystore, and also tear down the crypto stack.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-sync-engine/Source/UserSession/NSManagedObject+EncryptionContext.swift [hunk 1]
```diff
@@ -21,10 +21,12 @@ import WireDataModel
...
 
-    // TODO: [John] use flag here
-
     @objc
     public func deleteAndCreateNewEncryptionContext() {
-        zm_cryptKeyStore.deleteAndCreateNewBox()
+        proteusProvider.perform(
+            withProteusService: { _ in },
+            withKeyStore: { keyStore in keyStore.deleteAndCreateNewBox() }
+        )
```
Comment (johnxnguyen) [inline wire-ios-sync-engine/Source/UserSession/NSManagedObject+EncryptionContext.swift:29]: @typfel there's no core crypto api to clear the proteus data, should there be?

Comment (typfel) [reply wire-ios-sync-engine/Source/UserSession/NSManagedObject+EncryptionContext.swift:29]: There's no such API today, so you should would need delete underlaying sqlite store.

Comment (typfel) [reply wire-ios-sync-engine/Source/UserSession/NSManagedObject+EncryptionContext.swift:29]: Would it be useful to have such an API?

Comment (johnxnguyen) [reply wire-ios-sync-engine/Source/UserSession/NSManagedObject+EncryptionContext.swift:29]: I guess it's just to do some clean up, wouldn't it make sense from a security perspective to keep doing this?

Comment (typfel) [reply wire-ios-sync-engine/Source/UserSession/NSManagedObject+EncryptionContext.swift:29]: Yes, I think it makes sense to delete a defunct keystore, which anyway needs to do done if the user decide to login again.

But why do you need a CC API for this? You decide where CC stores the sqlite database. If it's stored in its own directory you can delete the whole directory and any auxiliary files will also be deleted.

Comment (johnxnguyen) [reply wire-ios-sync-engine/Source/UserSession/NSManagedObject+EncryptionContext.swift:29]: You're right, I was getting confused and thought that I needed to delete only the proteus keystore but keep the mls keystore intact, but we actually want to delete all of it.

**Review by David-Henner — APPROVED** [review]

**Review by agisilaos — APPROVED** [review]

**Pull request merged**